Update EIP-7801: Eip 7801 bitmask (#9058)
ethereum/EIPs: EIPS/eip-7801.md

---
eip: 7801
title: eth/70 - Sharded Blocks Protocol
description: Replaces block range with a bitmask representing 100,000-block spans in the handshake, committing to future block ranges

> author: Ahmad Bitar (@smartprogrammer93) <smartprogrammer@windowslive.com>, Giulio Rebuffo (@Giulio2002)

Reviewer: Please add yourself as author if you would like

## Abstract

> This EIP introduces a method enabling an Ethereum node to communicate its available block spans via a bitmask, where each bit represents a 100_000-block span within each 1_000_000 block range of chain history. Nodes use this bitmask to signal which spans of historical data they store and commit to storing future spans as they are created.  This enables peers to make informed decisions about data availability without first connecting and querying for availability. The bitmask repeats every 1 million blocks, making for a somewhat simple to reason about probability of data availability.

Reviewer: ```suggestion
This EIP introduces a method enabling an Ethereum node to communicate its available block spans via a bitmask, where each bit represents a 100_000-block span within each 1_000_000 block range of chain history. Nodes use this bitmask to signal which spans of historical data they store and commit to storing future spans as they are created. This enables peers to make informed decisions about data availability without first connecting and querying for availability. The bitmask repeats every 1 million blocks, making for a somewhat simple to reason about probability of data availability.
```

  - **Commitment to Future Ranges**: Participating nodes **MUST** retain 100_000 block spans which correspond to their advertised bitmask as blocks are added to the chain
  - **Bitmask Space**: The 100_000 range `blockBitmask` repeats every 1 million blocks, enabling efficient representation of historical data locality across epochs.

Upon connecting using `eth/70`, nodes exchange the `Status` message, including the `blockBitmask`. This single handshake message, which includes the bitmask, eliminates the need for additional message types.

Nodes should maintain connections regardless of a peer’s bitmask state, except when peer slots are full. In this case, nodes may prioritize connections with peers that serve more relevant shards.

If a node is interested in querying a specific block span, it can use the bitmask to determine which peers are more likely to have the data. If the peer they are connected to does not have the data, they can disconnect and connect to another peer, until they find what they have been looking for.

### ENR variation

> This EIP could be implemented without a new protocol version by adding a new field to the Ethereum Node Record (ENR) to advertise the bitmask. This would allow nodes to advertise their available block spans without requiring a handshake. However, this approach would not provide the same level of assurance as the handshake, as the ENR is not authenticated. Also, it would make the discovery process much simpler, as nodes just would need to take a look at the ENR to determine if a peer has the data they are looking for.

Reviewer: ```suggestion
This EIP could be implemented without a new protocol version by adding a new field to the Ethereum Node Record (ENR) to advertise the bitmask. This would allow nodes to advertise their available block spans without requiring a handshake. However, this approach would not provide the same level of assurance as the handshake, as the ENR is not authenticated. Also, it would make the discovery process much simpler, as nodes would only need to check the ENR to determine if a peer has the data they need.
```
